color-changer.js

```diff
@@ -0,0 +1,86 @@
+/*
+Write a Javascript module with the following specs:
+- When executed, it will change browser background to a random color every second.
+	// include a reference to this file in your html code to execute, using window.onload
+- Colors should not repeat.  (i.e. once a color is used, don’t use it again)
+	// verifies each hex value in the colorlist is unique using indexOf before pushing new colors
+- Stop changing the color after 10 seconds.
+	// iterates through a list of colors of given length (10) changing the background every second
+- Make it so someone can import the module into their own project
+	// code can be hosted and delivered via file server/CDN or the file can be downloaded from git
+*/
+
+// namespace
+var ColorChanger = (function () {
+
+	window.onload = function() {
+		initialize();
+		randomize();
+	}
+
+	// setup a display
+	function initialize(){
+		var colorContainer = document.createElement('div');
+		colorContainer.setAttribute('id', 'color-container');
+
+		var colorCodeText = document.createElement('h1');
+		colorCodeText.setAttribute('id', 'color-text');
+		colorCodeText.setAttribute('align', 'center');
+
+		document.body.appendChild(colorContainer);
+		document.getElementById('color-container').appendChild(colorCodeText);
+	}
+
+	/**
+	 * randomly generates a colorlist of hexidecimal characters
+	 * and changes the background color of the document.body
+	 * to the next color in the list every second
+	 */
+	function randomize() {
+		var colorList = _generateColorList(10);
+
+		colorList.forEach(function(color, index){
+			setTimeout(function(){
+				// change background color
+				document.body.style.backgroundColor = color;
+
+				// display current rgb value if exists
+				var displayDiv = document.getElementById('color-text');
+				if (displayDiv){
+					displayDiv.innerHTML = color;
+				}
+
+			}, index * 1000); // wait one second per iteration
+		});
+	}
+
+	// export for custom use
+	return {
+		initialize: initialize,
+		randomize: randomize
+	}
+
+	// helpers
+	function _generateColorList(length){
```

Comment:
What is the Big(0) runtime for this function? Can we change something to make it faster?

Reply:
As it is right now, Iterating through the colors array (which is of N size) and then calling indexOf within that loop would make the Big O =  O(n^2)  as I'm iterating through an array of N size and then checking an arrays elements (indexOf) of N size within that loop.  Which is O(n^2) Quadratic time.

I think I could optimize this by using a hashtable, generating a hex and storing it as the key if it is not already a key in the hashtable (using obj.hasOwnProperty or an if check such as if(!obj[key]){  //insert } which are both constant time Big O operations). I'd set the value of each key/value pair to true as well.
This would eliminate the indexOf check and bring the Big O to O(n), Linear time.